tests/unittests/Foundation/NSUndoManagerTests.mm

```diff
@@ -0,0 +1,718 @@
+    [self updateNumber:value];
+    [self.undoManager endUndoGrouping];
+}
+
+- (void)addToMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(removeFromMutableArray:) object:value];
+    [self.arrayOfNumbers addObject:value];
+}
+
+- (void)removeFromMutableArray:(NSNumber*)value {
+    [self.undoManager registerUndoWithTarget:self selector:@selector(addToMutableArray:) object:value];
+    [self.arrayOfNumbers removeObject:value];
+}
+
+@end
+
+TEST(NSUndoManager, BasicUndoTest) {
+    NSNumber* oldValue = [NSNumber numberWithInt:5];
+    NSNumber* newValue = [NSNumber numberWithInt:6];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:oldValue] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    [object updateNumber:newValue];
+
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, newValue.intValue);
+
+    [object.undoManager undo];
+
+    EXPECT_EQ(object.value.intValue, oldValue.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, MultipleUndoTests) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:val1] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    // Each undo operation here gets added to the single default undo group.
+    [object updateNumber:val2];
+    [object updateNumber:val3];
+    [object updateNumber:val4];
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    [object.undoManager undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    EXPECT_EQ([object.undoManager canUndo], NO);
+}
+
+TEST(NSUndoManager, undoRedoTests) {
+    NSNumber* val1 = [NSNumber numberWithInt:5];
+    NSNumber* val2 = [NSNumber numberWithInt:6];
+    NSNumber* val3 = [NSNumber numberWithInt:7];
+    NSNumber* val4 = [NSNumber numberWithInt:8];
+
+    UndoTestObject* object = [[[UndoTestObject alloc] initWithValue:val1] autorelease];
+    EXPECT_EQ([object.undoManager canUndo], NO);
+
+    [object updateNumber:val2];
+    [object updateNumber:val3];
+    [object updateNumber:val4];
+    EXPECT_EQ([object.undoManager canUndo], YES);
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    [object.undoManager undo];
+    EXPECT_EQ([object.undoManager canRedo], YES);
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    [object.undoManager redo];
+    EXPECT_EQ(object.value.intValue, val4.intValue);
+    EXPECT_EQ([object.undoManager canRedo], NO);
+    [object.undoManager undo];
+    EXPECT_EQ(object.value.intValue, val1.intValue);
+    EXPECT_EQ([object.undoManager canRedo], YES);
+}
+
+TEST(NSUndoManager, undoRedoTestsWithMutableObject) {
```

Comment:
Just a slightly different test of using undo operations to mutate an object instead of assigning it.

---
In reply to: [77097663](https://github.com/Microsoft/WinObjC/pull/900#discussion_r77097663) [](ancestors = 77097663)